### DISCUSSION TITLE
[Missing docs] Slugs and slugAttributes

### DISCUSSION THREAD

**Discussion by henze-housepedia** [area17/twill#860] [Q&A]
It seems straightforward, but when you need a concatenation of a string it is not. When you want to make a concatenation of an id and a title, you might think: "Hey, this $slugAttributes is an array of column names. I just do the following":

`public $slugAttributes = [
        'id',
        'title',
    ];`

Right? Wrong! the first is being used for the slug. The rest is used for dependencies. I don't know what this means, but it breaks if the columns aren't in the slugs table.
If you need a concatenation of two strings, just do the following:

`public $slugAttributes = [
        'slugifyThis',
    ];

    public function getSlugifyThisAttribute(): string
    {
        return $this->id.' '.$this->name;
    }`

 This does only work when not using translations. If you need translations to work, I suggest a rewrite of the getSlugParams. (create your own trait, use HasSlug {getSlugParams as public getTwillSlugParams;} )

This does nothing with the permalink, as this is being generated client side, and there are NO options to alter it's behavior, so advice is to turn permalinks off or deal with the fact that it does not show you what it is going to be.

As for the dependency thing, @ifox?

**Comment by Tofandel** [comment]
Bumping this issue

It's still not possible to have the permalink update correctly if there is multiple fields

The following create form with the previously mentionned workaround
```blade
<x-twill::input
        name="first_name"
        :label="twillTrans('speaker.first_name')"
        :translated="$translateTitle ?? false"
        :required="true"
        on-change="formatPermalink"
/>

<x-twill::input
        name="last_name"
        :label="twillTrans('speaker.last_name')"
        :translated="$translateTitle ?? false"
        :required="true"
        on-change="formatPermalink"
/>

@if ($permalink ?? true)
    <x-twill::input
            name="slug"
            :label="twillTrans('twill::lang.modal.permalink-field')"
            :translated="true"
            ref="permalink"
            :prefix="$permalinkPrefix ?? ''"
    />
@endif
```

![image](https://github.com/area17/twill/assets/6115458/2b67ac75-6315-46b6-98f2-04472b8a4ca6)

Any change to any field will trigger the permalink to be set to this field, we need to be able to send params to `formatPermalink` or provide a custom method

**Reply by Tofandel** [reply]
I was able to write a workaround like this but it's very limited

```blade
<x-twill::input
        name="first_name"
        :label="twillTrans('speaker.first_name')"
        :translated="$translateTitle ?? false"
        :required="true"
        on-change="formatPermalink($event + '-' + $el.querySelector('[name=last_name]').value)"
/>

<x-twill::input
        name="last_name"
        :label="twillTrans('speaker.last_name')"
        :translated="$translateTitle ?? false"
        :required="true"
        on-change="formatPermalink($el.querySelector('[name=first_name]').value + '-' + $event)"
/>

@if ($permalink ?? true)
    <x-twill::input
            name="slug"
            :label="twillTrans('twill::lang.modal.permalink-field')"
            :translated="true"
            ref="permalink"
            :prefix="$permalinkPrefix ?? ''"
    />
@endif
```

And adding a `create.blade.php` creates another issue
![image](https://github.com/area17/twill/assets/6115458/e5e8a6ed-429e-4e79-837b-bf551ea636d2)

Without the file the full name is correctly shown

**Reply by Tofandel** [reply]
Found a way to fix the missing title and reactivity issue, it's not pretty though

```blade
<x-twill::hidden
        :name="$titleFormKey"
        :translated="$translateTitle ?? false"
/>

<x-twill::input
        name="first_name"
        :label="twillTrans('speaker.first_name')"
        :translated="$translateTitle ?? false"
        :required="true"
        on-change="formatPermalink(TWILL.STORE.form.fields.find((f) => f.name === '{{$titleFormKey}}').value = $event + ' ' + $el.querySelector('[name=last_name]').value)"
/>

<x-twill::input
        name="last_name"
        :label="twillTrans('speaker.last_name')"
        :translated="$translateTitle ?? false"
        :required="true"
        on-change="formatPermalink(TWILL.STORE.form.fields.find((f) => f.name === '{{$titleFormKey}}').value = $el.querySelector('[name=first_name]').value + ' ' + $event)"
/>

@if ($permalink ?? true)
    <x-twill::input
            name="slug"
            :label="twillTrans('twill::lang.modal.permalink-field')"
            :translated="true"
            ref="permalink"
            :prefix="$permalinkPrefix ?? ''"
    />
@endif
```

I can absolutely PR support for multiple slug attributes later to never have to go through this again
